core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -172,8 +172,12 @@ private[kafka] object LogValidator {
         // Validate the timestamp
         validateTimestamp(record, now, messageTimestampType, messageTimestampDiffMaxMs)
         // Check if we need to overwrite offset, no in place assignment situation 3
-        if (logEntry.offset != expectedInnerOffset.getAndIncrement())
+        if (logEntry.offset != expectedInnerOffset.getAndIncrement()){
           inPlaceAssignment = false
+          if (record.magic() == messageFormatVersion) {
```

Comment:
This probably solves the common case, but I'm not sure it's a complete solution. We are only doing the validation for inner messages which do not match the expected offset. But the checksum might fail for any inner message. In fact, I think the issue is really that we are trusting the data prior to conversion. We should always validate the record checksum before doing any conversion. For example:

```scala
    records.deepEntries(true, BufferSupplier.NO_CACHING).asScala.foreach { logEntry =>
      val record = logEntry.record
      record.ensureValid()
      validateKey(record, compactedTopic)
...
```

Then we can remove `validatedRecords.foreach(_.ensureValid)` below. As far as I can tell, this is closer to how we do validation in trunk.

Reply:
> This probably solves the common case, but I'm not sure it's a complete solution. We are only doing the validation for inner messages which do not match the expected offset. But the checksum might fail for any inner message. In fact, I think the issue is really that we are trusting the data prior to conversion. We should always validate the record checksum before doing any conversion. For example:
> 
> ```scala
>     records.deepEntries(true, BufferSupplier.NO_CACHING).asScala.foreach { logEntry =>
>       val record = logEntry.record
>       record.ensureValid()
>       validateKey(record, compactedTopic)
> ...
> ```
> 
> Then we can remove `validatedRecords.foreach(_.ensureValid)` below. As far as I can tell, this is closer to how we do validation in trunk.

Thank you for your attention, I have revised a version based on your suggestion.

Reply:
> This probably solves the common case, but I'm not sure it's a complete solution. We are only doing the validation for inner messages which do not match the expected offset. But the checksum might fail for any inner message. In fact, I think the issue is really that we are trusting the data prior to conversion. We should always validate the record checksum before doing any conversion. For example:
> 
> ```scala
>     records.deepEntries(true, BufferSupplier.NO_CACHING).asScala.foreach { logEntry =>
>       val record = logEntry.record
>       record.ensureValid()
>       validateKey(record, compactedTopic)
> ...
> ```
> 
> Then we can remove `validatedRecords.foreach(_.ensureValid)` below. As far as I can tell, this is closer to how we do validation in trunk.

@hachikuji @guozhangwang

Reply:
I found a failure in the test case:

Reply:
> This probably solves the common case, but I'm not sure it's a complete solution. We are only doing the validation for inner messages which do not match the expected offset. But the checksum might fail for any inner message. In fact, I think the issue is really that we are trusting the data prior to conversion. We should always validate the record checksum before doing any conversion. For example:
> 
> ```scala
>     records.deepEntries(true, BufferSupplier.NO_CACHING).asScala.foreach { logEntry =>
>       val record = logEntry.record
>       record.ensureValid()
>       validateKey(record, compactedTopic)
> ...
> ```
> 
> Then we can remove `validatedRecords.foreach(_.ensureValid)` below. As far as I can tell, this is closer to how we do validation in trunk.

There is a failed in the test case, I found this is a test case bug, fixed in the trunk version: https://issues.apache.org/jira/browse/KAFKA-5063
Therefore, I understand that my modified version is actually no problem:
org.apache.kafka.streams.integration.ResetIntegrationTest > testReprocessingFromScratchAfterResetWithIntermediateUserTopic FAILED
    java.lang.AssertionError:
    Expected: <[KeyValue(3128935301215, 1), KeyValue(3128935301175, 1), KeyValue(3128935301195, 1), KeyValue(3128935301215, 1), KeyValue(3128935301235, 1), KeyValue(3128935301195, 1), KeyValue(3128935301215, 2), KeyValue(3128935301235, 1), KeyValue(3128935301255, 1), KeyValue(3128935301235, 2), KeyValue(3128935301275, 1), KeyValue(3128935301255, 1), KeyValue(3128935301255, 2), KeyValue(3128935301275, 1 ), KeyValue(3128935301295, 1), KeyValue(3128935301275, 2), KeyValue(3128935301315, 1), KeyValue(3128935301295, 1), KeyValue(3128935301275, 2), KeyValue(3128935301295, 2), KeyValue(3128935301315, 1) , KeyValue(3128935301275, 3), KeyValue(3128935301295, 2), KeyValue(3128935301315, 2), KeyValue(3128935301275, 3), KeyValue(3128935301295, 3), KeyValue(3128935301315, 2)]>
         But: was <[KeyValue(3128935301215, 1), KeyValue(3128935301175, 1), KeyValue(3128935301195, 1), KeyValue(3128935301215, 1), KeyValue(3128935301235, 1), KeyValue(3128935301195, 1), KeyValue(3128935301215 , 2), KeyValue(3128935301235, 1), KeyValue(3128935301255, 1), KeyValue(3128935301235, 2), KeyValue(3128935301275, 1), KeyValue(3128935301255, 1), KeyValue(3128935301255, 2), KeyValue(3128935301275, 1), KeyValue(3128935301295, 1), KeyValue(3128935301275, 2), KeyValue(3128935301315, 1), KeyValue(3128935301295, 1), KeyValue(3128935301275, 2), KeyValue(3128935301295, 2), KeyValue(3128935301315, 1 ), KeyValue(3128935301275, 3), KeyValue(3128935301295, 2), KeyValue(3128935301315, 2), KeyValue(3128935301275, 3), KeyValue(3128935301295, 3), KeyValue(3128935301315, 2), KeyValue(3128935301275, 4) , KeyValue(3128935301295, 3), KeyValue(3128935301315, 3)]>
        At org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
        At org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
        At org.apache.kafka.streams.integration.ResetIntegrationTest.testReprocessingFromScratchAfterResetWithIntermediateUserTopic(ResetIntegrationTest.java:193)

Reply:
> This probably solves the common case, but I'm not sure it's a complete solution. We are only doing the validation for inner messages which do not match the expected offset. But the checksum might fail for any inner message. In fact, I think the issue is really that we are trusting the data prior to conversion. We should always validate the record checksum before doing any conversion. For example:
> 
> ```scala
>     records.deepEntries(true, BufferSupplier.NO_CACHING).asScala.foreach { logEntry =>
>       val record = logEntry.record
>       record.ensureValid()
>       validateKey(record, compactedTopic)
> ...
> ```
> 
> Then we can remove `validatedRecords.foreach(_.ensureValid)` below. As far as I can tell, this is closer to how we do validation in trunk.

Thank you for your attention, I have revised a version based on your suggestion. @hachikuji  @guozhangwang

Reply:
There is a test case that didn't pass, but this is not caused by my modification. I found that the original kafka version could not pass this test case without any changes. This should be a test case bug:
org.apache.kafka.streams.integration.ResetIntegrationTest > testReprocessingFromScratchAfterResetWithIntermediateUserTopic FAILED
java.lang.AssertionError:
     Expected: <[KeyValue(3128960385575, 1), KeyValue(3128960385535, 1), KeyValue(3128960385555, 1), KeyValue(3128960385515, 1), KeyValue(3128960385575, 1), KeyValue(3128960385535, 1), KeyValue(3128960385555, 1), KeyValue (3128960385595, 1),

Reply:
Thank you very much for your sureness! Regarding kafka storage, I made some other optimization strategies based on the trunk branch.
https://issues.apache.org/jira/browse/KAFKA-8465

Reply:
Thank you very much for your sureness! Regarding kafka storage, I have made some other optimization strategies based on the trunk branch, hoping to get your affirmation.
https://issues.apache.org/jira/browse/KAFKA-8465       https://github.com/apache/kafka/pull/7133  @guozhangwang @hachikuji